### ISSUE TITLE
Paper 6/13 — Bounded Cache-Aware Expert Routing

### ISSUE THREAD

**Issue by murillo128** [murillo128/k3-out-of-core#115]
Parent: #109

**Editorial status:** `OUTLINE`

This issue coordinates the cache-aware routing method section only. It is not an execution contract.

## Purpose

Give the routing mechanism its own section. This is the paper's most distinctive algorithmic contribution and should not be buried inside the systems implementation.

### 5.1 Exact K3 expert selection

Define the exact reference decision clearly enough to establish what is and is not modified:

- K3 exposes 896 routed experts per routed layer;
- exact routing selects top-16;
- the paper keeps exact routing as the baseline/default;
- selected membership and the model's original expert weighting semantics must be treated separately.

Use the accepted K3 selection-score semantics from #77/#99 rather than a generic softmax/top-k description if K3's corrected selection differs materially.

### 5.2 Candidate expansion and residency-aware substitution

Conceptual algorithm:

```text
exact top-k selection
        ->
inspect a bounded top-M candidate set
        ->
identify selected experts whose service can be replaced
        ->
consider already-resident candidate experts
        ->
accept only substitutions satisfying the hard bounds
```

Make deterministic tie-breaking and contemporaneous residency explicit.

### 5.3 Hard per-swap regret and swap budget

Define the central bound using the exact K3 selection score, e.g.:

```text
regret(selected, candidate) = score(selected) - score(candidate)
```

and require:

```text
0 <= regret <= epsilon
realized substitutions <= max_swaps
```

Use mathematically precise notation in the manuscript once the exact score symbol is chosen.

The final frozen policy used for the main evidence is:

```text
S2_P50
candidate_count    = 32
max_swaps          = 2
max_score_regret   = 0.007303759455680847
```

### 5.4 Preserve K3 weighting semantics

State explicitly:

> The mechanism changes bounded expert membership, not the underlying unbiased expert weighting semantics.

This distinction is central to positioning relative to cache-prior reranking.

### 5.5 Policy selection and freeze

Explain enough of #77/#98 to establish that S2_P50 was selected before the later cross-prompt and long-horizon quality studies.

Relevant frozen #98 context:

```text
S2_P50 screening median TPS        0.505408936414
vs KNEE screening median TPS       +7.708%
loads/bytes per token vs KNEE      -15.866%
paired confirmations               all better; ~+6.8% to +7.1%
```

Keep #98's protocol distinction explicit; these values are policy-selection context and must not be pooled with #102 absolute physical results.

### 5.6 Why this is not a claim of semantic equivalence

Close the method section by stating that the local score bound constrains each intentional selection change but does not prove end-to-end semantic safety. Paper 9/13 evaluates that separately.

This is an important conceptual point because #99 later finds that simple cumulative/local regret summaries are weak predictors of long-horizon damage.

## Primary figure

A routing diagram should make the mechanism obvious:

```text
ranked router candidates

1  [selected exact]
2  [selected exact]
...
16 [selected exact]  ----\
17 [resident near-tie] <-- bounded substitution
18 [...]
...
32 [candidate boundary]

constraints:
  max_swaps <= 2
  per-swap regret <= epsilon
```

Show membership selection separately from final expert weighting.

## Prior-art positioning

Must cite and distinguish:

- Cache-Conditional Experts — training-free cache-aware membership/reranking with quality/locality frontier;
- MoE-ERAS — residency-aware selection balancing performance/accuracy;
- ReMoE — training-based router modification to increase reuse.

Safe novelty wording should emphasize the combination of:

- full Kimi K3 896->16 routing;
- deterministic hard per-swap regret + hard swap budget;
- training-free inference-time substitution;
- real subsequent route evolution rather than only frozen-route replay;
- physical NVMe-backed systems measurement;
- long-horizon direct-vs-free quality attribution.

Do **not** claim "first cache-aware MoE routing".

## Wording boundaries

- use `bounded` rather than `safe` unless referring to a precisely measured bound;
- do not equate router-score regret with semantic regret;
- do not describe S2_P50 as quality-preserving before task-level evidence;
- do not silently replace K3's exact scoring/weighting semantics with generic MoE notation where that would change the method.

**Comment by murillo128** [comment]
## Editorial update — Bounded Cache-Aware Expert Routing

Updated `paper/paper.md` on `paper/manuscript-v0` in commit `096562f49aa6105a402073acaff255f60db4c893`. Section 5 is now drafted and marked `EVIDENCE_CHECK`.

### Decisions frozen in the manuscript

- **Exact K3 semantics are explicit.** The section distinguishes the ordinary/unbiased router probability `p` from the corrected K3 selection score `s = p + exp_probs_b`; exact membership is the corrected-score top-16, while final expert weights continue to come from the ordinary probability tensor through the existing K3 gather/normalization/scale path.
- **Candidate expansion is bounded.** S2_P50 retains a top-32 envelope only as a membership decision surface; it still materializes exactly 16 unique experts.
- **Residency is read-only and contemporaneous.** A substitution is considered only when an unselected top-32 candidate is cheaper to service than a selected expert at the current layer boundary. Storage/provider state informs routing but does not choose experts independently.
- **Substitution semantics follow #77.** Eligible substitutions are deterministic and prioritized by avoided service cost, then lower selection-score regret, then original router rank, then stable expert-ID tie breaking. A replacement occupies the displaced top-16 slot and unaffected experts are not reordered.
- **Hard bounds are formalized.** For every accepted swap, `r = s(selected) - s(candidate)` must satisfy `0 <= r <= epsilon`; `max_swaps` is a hard per-routed-layer/per-token limit. S2_P50 is frozen at `candidate_count=32`, `max_swaps=2`, `max_score_regret=0.007303759455680847`. `max_score_regret=0` is explicitly treated as an exact-parity control, including ties.
- **No cumulative-regret control was retrofitted.** Cumulative regret remains analysis telemetry, not an adaptive runtime budget.
- **Original K3 weighting semantics are preserved.** Cache state can change bounded membership only; it is not inserted into router logits/probabilities or final expert weights, and the correction bias remains selection-only.

### Policy provenance / evidence boundary

- #77 owns the routing semantics, top-32 opportunity surface, and regret instrumentation.
- #98 owns physical policy selection. The manuscript records the protocol-specific reason S2_P50 was frozen: in the preregistered 21-cell `EXPLORATORY_MAX_SAFE` screening it was selected by the frozen rule, with +7.708% median TPS and -15.866% loads/bytes per token versus KNEE, followed by three paired confirmation gains of +6.832%, +6.857%, and +7.110%.
- Those #98 values are used only as **policy-selection context** and are explicitly not pooled with #102 absolute results.
- #102 and #99 occur after the policy freeze; neither is presented as retuning `M`, `max_swaps`, or `epsilon`.
- The local score bound is explicitly **not** a semantic-safety guarantee. The section points forward to #99/§8, where cumulative regret is only weakly predictive of long-horizon damage.

### Prior-art positioning

The section explicitly credits **Mixture of Cache-Conditional Experts** as direct training-free cache-aware membership prior art, **MoE-ERAS** as earlier residency-aware selection/performance–accuracy trade-off work, and **ReMoE** as training-based router fine-tuning for temporal reuse. The differentiation is deliberately narrow: K3 896→16, exact-route-relative top-32 substitution, hard per-swap selection-score regret, hard swap budget, unchanged K3 weighting semantics, real downstream route evolution, physical NVMe-backed evaluation, and later direct/free long-horizon quality attribution.

No `first cache-aware routing` claim is made. No new execution result, routing mechanism, threshold tuning, or quality claim was introduced by this editorial pass. No PR was created.
